File: app/src/main/java/com/github/anrimian/musicplayer/ui/common/images/glide/loaders/ExternalCompositionModelLoader.java
package com.github.anrimian.musicplayer.ui.common.images.glide.loaders;

import android.content.Context;
import android.graphics.Bitmap;

import androidx.annotation.NonNull;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.data.DataFetcher;
import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.data.models.composition.source.ExternalCompositionSource;
import com.github.anrimian.musicplayer.ui.common.images.glide.util.AppModelLoader;
import com.github.anrimian.musicplayer.ui.common.images.models.UriCompositionImage;
import com.github.anrimian.musicplayer.ui.utils.ImageUtils;

public class ExternalCompositionModelLoader extends AppModelLoader<UriCompositionImage, Bitmap> {

    private final Context context;

    public ExternalCompositionModelLoader(Context context) {
        this.context = context;
    }

    @Override
    protected Object getModelKey(UriCompositionImage uriCompositionImage) {
        return uriCompositionImage;
    }

    @Override
    protected void loadData(UriCompositionImage uriCompositionImage,
                            @NonNull Priority priority,
                            @NonNull DataFetcher.DataCallback<? super Bitmap> callback) {
        try {
            ExternalCompositionSource source = uriCompositionImage.getSource();
            byte[] imageBytes = source.getImageBytes();
            Bitmap bitmap = null;
            if (imageBytes != null) {
                int coverSize = context.getResources().getInteger(R.integer.icon_image_full_size);
                bitmap = ImageUtils.decodeBitmap(imageBytes, coverSize);
            }
            callback.onDataReady(bitmap);
        } catch (Exception e) {
            callback.onLoadFailed(e);
        }
    }

}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/common/images/glide/loaders/ExternalCompositionModelLoader.kt
package com.github.anrimian.musicplayer.ui.common.images.glide.loaders

import android.content.Context
import android.media.MediaMetadataRetriever
import com.bumptech.glide.Priority
import com.bumptech.glide.load.data.DataFetcher
import com.github.anrimian.musicplayer.ui.common.images.glide.util.AppModelLoader
import com.github.anrimian.musicplayer.ui.common.images.models.UriCompositionImage
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream

class ExternalCompositionModelLoader(
    private val context: Context,
) : AppModelLoader<UriCompositionImage, InputStream>() {

    override fun getModelKey(model: UriCompositionImage): Any {
        return model
    }

    override fun loadData(
        model: UriCompositionImage,
        priority: Priority,
        callback: DataFetcher.DataCallback<in InputStream>,
    ) {
        var mmr: MediaMetadataRetriever? = null
        try {
            mmr = MediaMetadataRetriever()
            mmr.setDataSource(context, model.uri)
            val imageBytes = mmr.embeddedPicture

            if (imageBytes != null) {
                callback.onDataReady(ByteArrayInputStream(imageBytes))
            } else {
                callback.onDataReady(null)
            }
        } catch (e: Exception) {
            callback.onLoadFailed(e)
        } finally {
            if (mmr != null) {
                try {
                    mmr.release()
                } catch (_: IOException) {}
            }
        }
    }

}
